core/src/services/gdrive/core.rs

```diff
@@ -0,0 +1,228 @@
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::collections::HashMap;
+use std::fmt::Debug;
+use std::fmt::Formatter;
+use std::sync::Arc;
+
+use crate::raw::HttpClient;
+use crate::Error;
+use crate::ErrorKind;
+use crate::Scheme;
+
+use http::request::Builder;
+use http::{header, Request, Response};
+use serde::Deserialize;
+use tokio::sync::Mutex;
+
+use crate::{
+    raw::{build_rooted_abs_path, new_request_build_error, AsyncBody, IncomingAsyncBody},
+    types::Result,
+};
+
+pub struct GdriveCore {
+    pub root: String,
+    pub access_token: String,
+    pub client: HttpClient,
+    pub path_cache: Arc<Mutex<HashMap<String, String>>>,
+}
+
+impl Debug for GdriveCore {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        let mut de = f.debug_struct("GdriveCore");
+        de.field("root", &self.root);
+        de.finish()
+    }
+}
+
+impl GdriveCore {
+    async fn get_abs_root_id(&self) -> Result<String> {
+        let root = "root";
+
+        if let Some(root_id) = self.path_cache.lock().await.get(root) {
+            return Ok(root_id.to_string());
+        }
+
+        let req = self
+            .sign(Request::get(
+                "https://www.googleapis.com/drive/v3/files/root",
+            ))
+            .body(AsyncBody::Empty)
+            .map_err(new_request_build_error)?;
+
+        let resp = self.client.send(req).await?;
+        let resp_body = &resp.into_body().bytes().await.map_err(|e| {
+            Error::new(ErrorKind::Unexpected, "read response body error")
+                .with_context("service", Scheme::Gdrive)
+                .set_source(e)
+        })?;
+
+        let gdrive_file: GdriveFile = serde_json::from_slice(resp_body).map_err(|e| {
+            Error::new(ErrorKind::Unexpected, "deserialize json error")
+                .with_context("service", Scheme::Gdrive)
+                .with_context("origin json value", String::from_utf8_lossy(resp_body))
+                .set_source(e)
+        })?;
+
+        let root_id = gdrive_file.id;
+
+        let mut cache_guard = self.path_cache.lock().await;
+        cache_guard.insert(root.to_owned(), root_id.clone());
+
+        Ok(root_id)
+    }
+
+    async fn get_file_id_by_path(&self, file_path: &str) -> Result<String> {
+        let path = build_rooted_abs_path(&self.root, file_path);
+
+        if let Some(file_id) = self.path_cache.lock().await.get(&path) {
+            return Ok(file_id.to_string());
+        }
+
+        let mut parent_id = self.get_abs_root_id().await?;
+        let file_path_items: Vec<&str> = path.split('/').filter(|&x| !x.is_empty()).collect();
+
+        for (i, item) in file_path_items.iter().enumerate() {
+            let mut query = format!(
+                "name = '{}' and parents = '{}' and trashed = false",
```

Comment:
This query seems wrong on this case:

```
root/a/a/a
```

Maybe we should get the parents one by one? Get `root/a`'s id frist and than query `item = "a" and parent = "<root/a's id>"`.

Reply:
I tried `op.read("root/a/a/a")`, it works. Maybe you think there is wrong with `https://www.googleapis.com/drive/v3/files/root` in `get_abs_root_id`? `root` in the query string isn't folder or file name, it can be see as a file id. see [here](https://developers.google.com/drive/api/reference/rest/v3/files/get?apix_params=%7B%22fileId%22%3A%2211cQ4X7viHDPWMTaObm638_rovNOGznSQ%22%7D#http-request).

Reply:
Ok, I got your code. 

So we can do it more faster by checking `root/a/a/`, `root/a/`, `root/` in order to find the closest parent so that we don't need to get fileid from root every time.

Think about this case `/root/a/a/a`, `/root/a/a/b`.

Reply:
will consider in next PR